### PR TITLE
Feature/refactor driver

### PR DESCRIPTION
**Pull request by eschizoid**

### CODE REVIEW

**Comment by eschizoid** [comment]
@argenisleon Almost done with the ```IO``` package refactor, will start adding tests as my next step. However, feel free to start adding comments 😃

**Comment by argenisleon** [comment]
@eschizoid I am on it :)!

**Comment by FavioVazquez** [comment]
@eschizoid  there are some trailing whitespaces in:

optimus/io/oracle.py line 18
optimus/io/postgres.py line 27
optimus/io/redshift.py line 27

can you fix that please? For more information visit:

 https://app.codacy.com/manual/Iron-AI/Optimus/pullRequest?prid=4388002

**Comment by eschizoid** [comment]
@FavioVazquez Done! This weekend I will push a good set of unit tests :)

**Comment by eschizoid** [comment]
@FavioVazquez @argenisleon I was also thinking that maybe all the databases drivers should be within the scope of the ```factory.py```

```
class DriverFactory:

    def get(driver_type) -> AbstractDriver:
     
    class OracleDriver:
    ...
     class PostgresDriver:

     elif driver_type == DriverProperties.ORACLE.__str__():
            return OracleDriver()

```

**Comment by argenisleon** [comment]
@eschizoid sound good for me :)

**Comment by eschizoid** [comment]
@FavioVazquez @argenisleon I was able to manually test (manually) the following databases:

- [ ]  Postgres

**Comment by eschizoid** [comment]
Hey @argenisleon wondering if you had time to review this bad boy.

**Comment by argenisleon** [comment]
@eschizoid I'm on it!

File: optimus/io/abstract_driver.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+        Returns the query for all tables in the database.
+        :param query_parameters: query parameters
+        :return: a query for all the tables in a given database
+        """
+        pass
+
+    @abstractmethod
+    def table_name_query(self, *args, **kwargs) -> str:
+        """
+        Returns the query for all tables in the database.
```
Comment (argenisleon) [inline optimus/io/abstract_driver.py]: The description is the same that ```def table_names_query```. Is this correct?

File: optimus/io/properties.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    PRESTO = {
+        "name": "presto",
+        "port": 8080,
+        "java_class": "com.facebook.presto.jdbc.PrestoDriver",
+        "table_name": ""
+    }
+    REDSHIFT = {
+        "name": "redshift",
+        "port": 5439,
+        "java_class": "",
```
Comment (argenisleon) [inline optimus/io/properties.py]: Should we use ```com.amazon.redshift.jdbc.Driver``` for the java class?

File: optimus/optimus.py [hunk 3]
```diff
@@ -171,14 +171,14 @@ def __init__(self, session=None, master="local[*]", app_name="optimus", checkpoi
 
     @staticmethod
     def connect(driver=None, host=None, database=None, user=None, password=None, port=None, schema="public",
-                oracle_tns=None, oracle_service_name=None, oracle_sid=None, presto_catalog=None,
-                cassandra_keyspace=None, cassandra_table=None):
+                oracle_service_name=None, oracle_sid=None, presto_catalog=None, cassandra_keyspace=None,
```
Comment (argenisleon) [inline optimus/optimus.py]: What happened to the oracle_tns param?

File: optimus/io/properties.py (hunk 2 above)
Comment (eschizoid) [reply optimus/io/properties.py]: Currently we don't set that option for Redshift (line 154 jdbc.py):

```
# Driver
if self.db_driver == DriverProperties.ORACLE.value["name"] \
        or self.db_driver == DriverProperties.POSTGRESQL.value["name"] \
        or self.db_driver == DriverProperties.PRESTO.value["name"]:
    conf.option("driver", self.driver_option)```

We could definitely add it though, let me know your thoughts (unfortunately now sure how to really test Redshift)

File: optimus/optimus.py (hunk 3 above)
Comment (eschizoid) [reply optimus/optimus.py]: None of the connection strings are currently using it:

None of the connection strings use ```oracle_tns```, they use either ```oracle_service_name or ```oracle_sid``` (see oracle.py line 17)

```
return f"""jdbc:{conn_properties["driver"]}:thin:@{conn_properties["host"]}:{conn_properties["port"]}/{conn_properties["oracle_sid"]}""" \
            if conn_properties["oracle_sid"] is not None \
            else f"""jdbc:{conn_properties["driver"]}:thin:@//{conn_properties["host"]}:{conn_properties["port"]}/{conn_properties["oracle_service_name"]}"""
```
I have successfully test oracle using both connectuion strings and they work :)

File: optimus/io/abstract_driver.py (hunk 1 above)
Comment (eschizoid) [reply optimus/io/abstract_driver.py]: Good catch, fixed

File: optimus/optimus.py (hunk 3 above)
Comment (argenisleon) [reply optimus/optimus.py]: @eschizoid I created https://hub.docker.com/repository/docker/ironmussa/optimus, Do you have a Docker ID so I can add you to the team?

**Comment by eschizoid** [comment]
Oracle tests added, let me know what you guys think

File: optimus/optimus.py (hunk 3 above)
Comment (eschizoid) [reply optimus/optimus.py]: > @eschizoid I created https://hub.docker.com/repository/docker/ironmussa/optimus, Do you have a Docker ID so I can add you to the team?

eschizoid

Comment (argenisleon) [reply optimus/optimus.py]: Added!

**Comment by eschizoid** [comment]
@argenisleon :(

```
docker push ironmussa/test-db:oracle-19.3.0-ee
The push refers to repository [docker.io/ironmussa/test-db]
bdda7e722ca2: Preparing
22f317703c61: Preparing
1aec78830b5b: Preparing
f07908bedb2f: Waiting
77144599f113: Waiting
4fee40bcfecf: Waiting
denied: requested access to the resource is denied
```

**Comment by argenisleon** [comment]
@eschizoid 
Please try to push to https://hub.docker.com/repository/docker/ironmussa/optimus

File: optimus/io/cassandra.py [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
+
+@singleton
+class CassandraDriver(AbstractDriver):
+    """Cassandra Database"""
+
+    def properties(self) -> Enum:
+        return DriverProperties.CASSANDRA
+
+    def url(self, *args, **kwarg) -> str:
```
Comment (argenisleon) [inline optimus/io/cassandra.py]: Missing a 's' in 'kwarg' in all functions

**Comment by eschizoid** [comment]
@argenisleon image pushed and ```docker-compose.yml``` updated
Let me know if you want me to remove the Oracle ```Dockerfile```

**Comment by eschizoid** [comment]
@argenisleon All PR comments have been addressed and merge conflicts have been resolved.

Let me know if you have more comments.

Meanwhile, I will start working on unit testing the IO package. However, I would like to merge this bad boy since keeping up to date the work for a solid month can be tedious.

**Comment by argenisleon** [comment]
@eschizoid I agree. I'll start working on the merge now.

**Pull request merged**